### ISSUE TITLE
How to do parallelization on GPU?

### ISSUE THREAD

**Issue by amyumhna** [facebookresearch/faiss#2070]
# Summary

The original data is a 3D tensor, and the training and distance calculation is done by extracting two dimensions.
On the GPU, I want to run different Faisses concurrently.

On the CPU, it is possible to use `concurrent.futures.ThreadPoolExecutor`

# Platform

OS: Ubuntu20.04

Faiss version: faiss 1.5.3

Installed from: pip

Faiss compilation options: 

Running on:
- [ ] CPU
- [x] GPU

Interface: 
- [ ] C++
- [x] Python

# Reproduction instructions

```

import torch
import faiss
import faiss.contrib.torch_util

train_data = torch.rand((dim,train_num,row_num) , device=0)
query_data = torch.rand((dim,query_num,row_num) , device=0)
k = 5

def calc_Faiss(row):
    
    xb = train_data[:,:,row]
    xq = query_data[:,:,row]
    res = faiss.StandardGpuResources()
    flat_config = faiss.GpuIndexFlatConfig()
    index = faiss.GpuIndexFlatL2(res, dim, flat_config) 
    index.add(xb)
    D, I = index.search(xq, k)
    return D

## I want to parallelize this iteration.
for row in range(row_num):
    calc_Faiss(row)

```

**Comment by mdouze** [comment]
The sequential version of this code does not work, does it? The dimensions of xq and xb are transposed w.r.t. what they should be. 
Concurrent execution on a single GPU is not supported, see https://github.com/facebookresearch/faiss/wiki/Threads-and-asynchronous-calls#thread-safety
If dim, train_num, and query_num are large enough then the GPU should be running at close to peak performance anyways, so there is nothing to gain from parallelization. When either of the dimensions is small, then there may be more efficient approaches.
